src/org/sipdroid/sipua/PublishAgent.java

```diff
@@ -0,0 +1,213 @@
+		this.publish(BasicStatus.OPEN, SipStack.default_expires, "");
+	}
+
+	public void publish(BasicStatus status, int expireTime, String note) {
+		SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
+		Boolean publish_enable_status = prefs.getBoolean("publish_enable", true);
+		if (publish_enable_status == true) {
+			MessageDigest md = null;
+			String tupleId;
+			try {
+				md = MessageDigest.getInstance("MD5");
+				tupleId =  md.digest(user_profile.username.getBytes()).toString();
+			} catch (NoSuchAlgorithmException e) {
+				tupleId = user_profile.username;
+				e.printStackTrace();
+			}
+			String from = user_profile.username;
+			String entity = "sip:" + user_profile.username;
+			String xml = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>" +
+							"<presence xmlns=\"urn:ietf:params:xml:ns:pidf\"" +
+							" entity=\"" + entity + "\">" +
+							"<tuple id=\"" + tupleId + "\">" +
+							"<status>" +
+							"<basic>" + status.pidf() + "</basic>" +
+							"<note>" + note + "</note>" +
+							"</status>" +
+							"</tuple>" +
+							"</presence>";
+			MessageFactory msgf = new MessageFactory();
+			Message req = msgf.createPublishRequest(sip_provider, new NameAddress(from), "presence", expireTime, "application/pidf+xml", xml);
+			TransactionClient t = new TransactionClient(sip_provider, req, this);
+			t.request();
+		}
+	}
+
+	public void unPublish() {
+		SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
+		String from = user_profile.username;
+		Boolean publish_enable_status = prefs.getBoolean("publish_enable", true);
+		if (publish_enable_status == true) {
+			MessageFactory msgf = new MessageFactory();
+			Message req = msgf.createPublishRequest(sip_provider, new NameAddress(from), "presence", 0, null, null);
+			TransactionClient t = new TransactionClient(sip_provider, req, this);
+			t.request();
+		}
+	}
+
+	private boolean generateRequestWithProxyAuthorizationheader(Message resp, Message req) {
+		ProxyAuthenticateHeader pah = resp.getProxyAuthenticateHeader();
+		ProxyAuthorizationHeader ah1 = (new DigestAuthentication(
+				req.getTransactionMethod(), req.getRequestLine().getAddress()
+				.toString(), pah, null, null, username, passwd))
+				.getProxyAuthorizationHeader();
+		req.setProxyAuthorizationHeader(ah1);
+		return true;
+	}
+
+	public void onTransSuccessResponse(TransactionClient tc, Message resp) {
+		StatusLine status = resp.getStatusLine();
+		String log = MessageFormat.format("onTransSuccessResponse: status {0}: {1}", status.getCode(), status);
+		logger.fine(log);
+	}
+
+	public void onTransFailureResponse(TransactionClient tc, Message resp) {
+		StatusLine status = resp.getStatusLine();
+		int code = status.getCode();
+		if (code == 401 || code == 407) {
+			processAuthenticationResponse(tc, resp, code);
+		}
+		String log = MessageFormat.format("onTransFailureResponse: status {0}: {1}", code, status);
+		logger.fine(log);

```

Comment:
it would be better to have this logger stuff before the ```if``` statement, logging what happens before any logic tries to process it